### PR TITLE
Released Neural Networks Compression Framework

### PR DESCRIPTION
**Pull request by AlexKoff88**
## Key features:

- Support of quantization and sparsity algorithms with fine-tuning.
- Automatic model graph transformation. The model is wrapped by the custom class and additional layers are inserted in the graph. The transformations are configurable.
- Common interface for compression methods.
- GPU-accelerated layers for fast model fine-tuning.
- Distributed training support.
- Configuration file examples for sparsity, quantization and sparsity with quantization. Each type of compression requires only one additional fine-tuning stage.
- Export models to ONNX format that is supported by [OpenVINO](https://github.com/opencv/dldt) Toolkit.

### CODE REVIEW

File: pytorch_toolkit/nncf/nncf/README.md [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+### Non-structured sparsity
+Sparsity algorithm zeros weights in convolutional and fully-connected layers in a non-structured way, so that zero values are randomly distributed inside the tensor. Most of the sparsity algorithms set less to zero the less important weights but the criteria of how they do it is different. The framework contains several implementations of sparsity methods.
+
+#### RB-Sparsity
+
+This section describes the RB-Sparsity (Regularization-Based) algorithm that is implemented in this framework. The method is based on $`L_0`$ regularization that encourages parameters of the model to become zero:
+
+```math
+    ||\theta||_0 = \sum_{i=0}^{|\theta|} \lbrack \theta_i = 0 \rbrack
```
Comment (AlexanderDokuchaev) [inline pytorch_toolkit/nncf/nncf/README.md]: It's not works on Github.

Use https://www.codecogs.com/latex/eqneditor.php to render latex.
``

File: pytorch_toolkit/nncf/nncf/README.md [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+```
+
+However, since the $`L_0`$ norm is non-differentiable, we relax it by adding multiplicative noise to the model parameters:
+
+```math
+    \theta_{sparse}^{(i)} = \theta_i \cdot \epsilon_i, \quad
+    \epsilon_i \sim \mathcal{B}(p_i)
+```
+
+Here $`\epsilon_i`$ may be interpreted as a binary mask that selects which weights should be zeroed, hence we add the regularizing term to the objective function that encorages desired level of sparsity to our model:
```
Comment (AlexanderDokuchaev) [inline pytorch_toolkit/nncf/nncf/README.md]: encorages -> encourages

File: pytorch_toolkit/nncf/examples/classification/README.md [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+`python main.py -m test --config=configs/quantization/mobilenetV2_imagenet_int8.json --resume=../../results/quantization/mobilenetV2_int8/6/checkpoints/epoch_1.pth --to-onnx=../../results/mobilenetV2_int8.onnx`
+
+#### Export to OpenVINO Intermediate Representation (IR)
+
+To export a model to OpenVINO IR and run it using Intel Deep Learning Deployment Toolkit please refer to this [tutorial](https://software.intel.com/en-us/openvino-toolkit).
+
+### Results   for INT8 quantization
+
+| Model | Dataset | FP32 baseline | Compressed model accuracy | Config path | Checkpoint |
+|--|--|--|--|--|--|
```
[the remark this answers is not in the document]
Comment (AlexKoff88) [reply pytorch_toolkit/nncf/examples/classification/README.md]: Done

File: pytorch_toolkit/nncf/docs/Configuration.md [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+Logically all parameters are divided into two groups:
+- **Training parameters** that are related to the training process and used by training sample (e.g. *learning rate*, *weight decay*, *optimizer type*, etc.)
+- **Compression parameters** that are related to the compression method (e.g. *compression algorithm name*, *target sparsity ratio*, etc.)
+
+**Note**: Compression parameters are optional and can be skipped. In this case, a regular model training will be executed without any compression method.
+
+## Example of the configuration file
+This is an example of the configuration file that contains training and compression parameters for two algorithms: sparsity and INT8 quantization which are applied to MobileNet v2 model on ImageNet classification task.  See the description of each setting as comments inside the file.
+
+```json
```
[the remark this answers is not in the document]
Comment (AlexKoff88) [reply pytorch_toolkit/nncf/docs/Configuration.md]: Done

File: pytorch_toolkit/nncf/docs/Algorithm.md [hunk 5]
```diff
@@ -0,0 +1,39 @@
+# Compression Method API
+
+This is a description of the Compression Method API. The Neural Network Compression Framework (NNCF) is designed to be extendable and it is assumed that the end user can easily add new compression methods to the framework implementing an API which is specified [here](../common.py).
+The compression method is divided into three logical parts:
+- The algorithm itself that implements API of `CompressionAlgorithm` class and logic of compression method.
+- The scheduler that implements API of `CompressionScheduler` class and compression method scheduling control logic.
+- The loss that implements API of `CompressionLoss` class and loss function that is used in the compression method during the training process.
+
+Note: In case a custom compression method does not implement its own scheduler and loss then the default implementations are used.
```
[the remark this answers is not in the document]
Comment (AlexKoff88) [reply pytorch_toolkit/nncf/docs/Algorithm.md]: Done

File: pytorch_toolkit/nncf/ReleaseNotes.md [hunk 6]
```diff
@@ -0,0 +1,18 @@
+# Release Notes
+
+## Introduction
+*Neural Network Compression Framework  (NNCF)* is a toolset for Neural Networks model compression. The framework organized as a Python module that can be built and used standalone or within samples distributed with the code.  The samples demonstrate the usage of compression methods on public models and datasets for three different use cases: Image Classification, Object Detection, and Semantic Segmentation.
```
[the remark this answers is not in the document]
Comment (AlexKoff88) [reply pytorch_toolkit/nncf/ReleaseNotes.md]: Done

File: pytorch_toolkit/nncf/nncf/README.md [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+
+```math
+s = \sigma^{-1}(p) = log (\frac{p}{1 - p})
+```
+and reparametrize sampling of $`\epsilon`$ as:
+
+```math
+    \epsilon = \lbrack \sigma(s + \sigma^{-1}(\xi)) > \frac{1}{2} \rbrack, \quad \xi \sim \mathcal{U}(0,1)
+```
+it can be shown that with this reparametrization, probablity of keeping a particular weight during the forward pass is exatcly $`\mathbb{P}( \epsilon_i = 1) = p_i`$. We only use weights with $`p_i > \frac{1}{2}`$ at test time. To make the objective function differentiable, we treat threshold function $`t(x) = [x > c]`$ as straight through estimator i.e. $`\frac{d t}{dx} = 1`$
```
[the remark this answers is not in the document]
Comment (AlexKoff88) [reply pytorch_toolkit/nncf/nncf/README.md]: Done

File: pytorch_toolkit/nncf/nncf/README.md (hunk 1 above)
Comment (AlexKoff88) [reply pytorch_toolkit/nncf/nncf/README.md]: Done

File: pytorch_toolkit/nncf/nncf/README.md (hunk 2 above)
Comment (AlexKoff88) [reply pytorch_toolkit/nncf/nncf/README.md]: Done

File: pytorch_toolkit/nncf/examples/classification/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (AlexKoff88) [reply pytorch_toolkit/nncf/examples/classification/README.md]: Done

File: pytorch_toolkit/nncf/nncf/README.md (hunk 1 above)
Comment (AlexKoff88) [reply pytorch_toolkit/nncf/nncf/README.md]: Done
